src/query/phrase_query/exact_phrase_scorer.rs

```diff
@@ -0,0 +1,259 @@
+        if self.scoring_enabled {
+            let intersection_len = self.compute_phrase_match();
+            let count = intersection_count(&self.left[..intersection_len], &self.right[..]) as u32;
+            self.phrase_count = count;
+            count > 0u32
+        } else {
+            self.phrase_exists()
+        }
+    }
+
+    fn phrase_exists(&mut self) -> bool {
+        let intersection_len = self.compute_phrase_match();
+        intersection_exists(&self.left[..intersection_len], &self.right[..])
+    }
+}
+
+impl<TPostings: Postings> ExactPhraseScorer<TPostings> {
+    pub fn new(
+        term_postings: Vec<(usize, TPostings)>,
+        similarity_weight: Bm25Weight,
+        fieldnorm_reader: FieldNormReader,
+        scoring_enabled: bool,
+    ) -> ExactPhraseScorer<TPostings> {
+        let num_terms = term_postings.len();
+        let mut scorer = ExactPhraseScorer {
+            intersection_docset: Intersection::new(get_postings_with_offset(term_postings)),
+            num_terms,
+            left: Vec::with_capacity(100),
+            right: Vec::with_capacity(100),
+            phrase_count: 0u32,
+            similarity_weight,
+            fieldnorm_reader,
+            scoring_enabled,
+        };
+        if scorer.doc() != TERMINATED && !scorer.phrase_match() {
+            scorer.advance();
+        }
+        scorer
+    }
+
+    fn compute_phrase_match(&mut self) -> usize {
+        {
+            self.intersection_docset
+                .docset_mut_specialized(0)
+                .positions(&mut self.left);
+        }
+        let mut intersection_len = self.left.len();
+        for i in 1..self.num_terms - 1 {
+            {
+                self.intersection_docset
+                    .docset_mut_specialized(i)
+                    .positions(&mut self.right);
+            }
+            intersection_len = intersection(&mut self.left[..intersection_len], &self.right[..]);
+            if intersection_len == 0 {
+                return 0;
+            }
+        }
+        self.intersection_docset
+            .docset_mut_specialized(self.num_terms - 1)
+            .positions(&mut self.right);
+        intersection_len
+    }
+}
+
+impl<TPostings: Postings> DocSet for ExactPhraseScorer<TPostings> {
+    fn advance(&mut self) -> DocId {
+        loop {
+            let doc = self.intersection_docset.advance();
+            if doc == TERMINATED || self.phrase_match() {
+                return doc;
+            }
+        }
+    }
+
+    fn seek(&mut self, target: DocId) -> DocId {
+        debug_assert!(target >= self.doc());
+        let doc = self.intersection_docset.seek(target);
+        if doc == TERMINATED || self.phrase_match() {
+            return doc;
+        }
+        self.advance()
+    }
+
+    fn doc(&self) -> DocId {
+        self.intersection_docset.doc()
+    }
+
+    fn size_hint(&self) -> u32 {
+        self.intersection_docset.size_hint()
+    }
+}
```

Comment:
Both structs have the exact same impl for `DocSet` but there's allready a blanket implem for `DocSet`. Unsure how to workaround this.